src/pudl/analysis/allocate_net_gen.py

```diff
@@ -235,25 +195,101 @@ def allocate_gen_fuel_by_gen_pm_fuel(gf, gen, gens, drop_interim_cols=True):
                 "fuel_consumed_mmbtu",
             ]
         ]
+    _test_gen_fuel_allocation(gen, gen_pm_fuel_frac)
     return gen_pm_fuel_frac
 
 
+def aggregate_generation_fuel_by_generator(
+    pudl_out, gen_pm_fuel: pd.DataFrame
+) -> pd.DataFrame:
+    """
+    Aggregate gen fuel data columns to generators.
+
+    The generation_fuel_eia923 table includes net generation and fuel
+    consumption data at the plant/fuel type/prime mover level. The most
+    granular level of plants that PUDL typically uses is at the plant/generator
+    level. This function takes the plant/fuel type/prime mover level allocation,
+    aggregates it to the generator level and then denormalizes it to make it
+    more structurally in-line with the original generation_eia923 table.
+
+    Args:
+        pudl_out (pudl.output.pudltabl.PudlTabl): An object used to create the
```

Comment:
You can put this type in the type hints up in the function call signature instead, so they're all in the same place.

Reply:
i'm getting circular references if i import PudlTabl into this module